e2e/pages/vm-sidebar.po.ts

```diff
@@ -171,4 +172,142 @@ export class VMSidebar extends CloudstackUiPage {
+      ),
+    ).click();
+  }
+
+  getSelectedSO() {
+    return element(
+      by.xpath(
+        "//mat-radio-button[@class='mat-radio-button mat-accent mat-radio-checked']/ancestor::td/preceding-sibling::td//span",
+      ),
+    ).getText();
+  }
+
+  getChangedSO() {
+    return element(
+      by.xpath(`//cs-service-offering-details//div[text()="Name"]/following-sibling::div`),
+    ).getText();
+  }
+
+  checkWarning() {
+    element(by.css('.message.warning.ng-star-inserted')).isPresent();
+    return element(by.css('.message.warning.ng-star-inserted')).getText();
+  }
+
+  isEnabledChangeSO() {
+    expect(
+      element(by.css('cs-service-offering-details button[ng-reflect-disabled=true]')),
+    ).toBeTruthy('Change SO button is enabled');
+  }
+
+  isEnabledAddAffinityGroup() {
+    expect(element(by.css('cs-affinity-group button[ng-reflect-disabled=true]'))).toBeTruthy(
+      'Affinity group button is enabled',
+    );
+  }
+
+  isEnabledAddSSHKey() {
+    expect(element(by.css('cs-vm-ssh-keypair button[ng-reflect-disabled=true]'))).toBeTruthy(
+      'Add SHH Key is enabled',
+    );
+  }
+
+  clickChangeSSHKey() {
+    const edit = element(by.css('cs-vm-ssh-keypair button'));
+    const EC = protractor.ExpectedConditions;
+    browser.wait(EC.elementToBeClickable(edit), 2000, 'No edit Group button').then(() => {
+      edit.click();
+      this.waitDialogModal();
+    });
+  }
+
+  selectSSHKey(index) {
+    element(by.css('cs-ssh-keypair-reset mat-select')).click();
+    element
+      .all(by.css('mat-option'))
+      .get(index)
+      .click();
+    return element
+      .all(by.css('mat-option span'))
+      .get(index)
+      .getText();
+  }
+
+  getSHHKey() {
+    return element(by.css('cs-vm-ssh-keypair span')).getText();
+  }
+
+  checkResetWarning() {
+    return browser.wait(
+      protractor.ExpectedConditions.presenceOf(element(by.css('.mat-dialog-content.notification'))),
+      5000,
+    );
+  }
+
+  isEnabledCreateAffGroupButton() {
+    return element(by.css('.add-rule-button.mat-icon-button')).isEnabled();
+  }
+
+  setIncorrectAffGroupName(name) {
+    element(by.css('input[formControlName=name]')).sendKeys(name);
+  }
+
+  getErrorText() {
+    element(by.css('cs-affinity-group-selector mat-card-header')).click();
+    const EC = browser.ExpectedConditions;
+    browser.wait(
+      EC.presenceOf(element(by.css('.mat-error.ng-star-inserted'))),
+      5000,
+      'Error message is not present',
+    );
+    return element(by.css('.mat-error.ng-star-inserted')).getText();
+  }
+
+  clickDeleteAffGroup() {
+    const del = element(by.css('cs-affinity-group .row.ng-star-inserted button'));
+    const EC = protractor.ExpectedConditions;
+    browser.wait(EC.elementToBeClickable(del), 2000, 'does not clicable button').then(() => {
+      del.click();
+    });
+  }
```

Comment:
Similar to the comment for the selection of SO. Remove waiting if it is not needed.